ykllvmwrap/src/jitmodbuilder.cc

```diff
@@ -187,27 +187,25 @@ class JITModBuilder {
     this->FAddrKeys = FAddrKeys;
     this->FAddrVals = FAddrVals;
     this->FAddrLen = FAddrLen;
+
+    JITMod = new Module("", AOTMod->getContext());
   }
 
   // FIXME: this function needs to be refactored.
   // https://github.com/ykjit/yk/issues/385
   Module *createModule() {
-    LLVMContext &JITContext = AOTMod->getContext();
-    JITMod = new Module("", JITContext);
-    uint64_t TraceIdx = NextTraceIdx.fetch_add(1);
-    if (TraceIdx == numeric_limits<uint64_t>::max())
-      errx(EXIT_FAILURE, "trace index counter overflowed");
-
+    LLVMContext &JITContext = JITMod->getContext();
     // Get var args from start_tracing call.
     auto Inputs = getTraceInputs(AOTMod->getFunction(FuncNames[0]), BBs[0]);
 
-    std::vector<Type *> InputTypes;
-    for (auto Val : Inputs) {
-      InputTypes.push_back(Val->getType());
-    }
-
     // Create function to store compiled trace.
+    uint64_t TraceIdx = NextTraceIdx.fetch_add(1);
+    if (TraceIdx == numeric_limits<uint64_t>::max())
```

Comment:
*Technically* this check is, I think, wrong: the trace index will overflow on the next `fetch_add`. But... is the explicit failure worth it? In other words, just before line 202, I'd add `assert(NextTraceIdx != uin64_t::max())` (or something to that effect).

Reply:
This is existing code that has been moved. Shall we fix this in a separate commit?

Reply:
OK. There are a couple of these, so doing it in a new commit (but as part of this PR) is fine with me.

Reply:
Wait. The fix you are proposing wouldn't be thread-safe, would it? The check itself needs to be atomic, which is why I think I tried to do it with a single `fetch_add`.

You are right that we lose one trace ID though...

Reply:
Is this `fetch_add` an atomic instruction? It doesn't have an ordering, so I assumed it's just a bog-standard non-thread-safe version.

Reply:
It's using the default ordering, which is actually too strong here I think:
```
T fetch_add( T arg,
             std::memory_order order = std::memory_order_seq_cst ) noexcept;
```

Since the counter only depends on itself, and since there's a total ordering on atomic ops for the same value, I think it can be weakened to `memory_order_relaxed`.

Do you agree?

Reply:
Yes, relaxed ordering is fine. I suggest this should actually be in a self contained function called something like `next_trace_idx` that can encapsulate this check. It should probably be something like (in Rust, because I'm lazy):

```rust
let x = self.current_idx.fetch_add(Ordering::Relaxed, 1);
debug_assert!(x < u64::max());
x
```

[We could squeeze out the last value in the u64 range... but it doesn't really seem worth it.]
```

Reply:
That's pretty much what we have. `x < u64::max()` is equivalent to checking that we don't have `x == u64::max`.

I will move it into a function though. Good idea.